misc-utils/enosys.c

```diff
@@ -100,13 +106,61 @@ static void __attribute__((__noreturn__)) usage(void)
 struct blocked_number {
 	struct list_head head;
 	long number;
+	int ret;
 };
 
+static struct blocked_number *parse_block(const char *s, int ret, const struct syscall entities[], size_t n_entities)
+{
+	struct blocked_number *blocked;
+	const char *name, *error_name;
+	long blocked_number;
+	char *colon;
+	bool found;
+	size_t i;
+
+	colon = strchr(s, ':');
+	if (colon) {
+		name = xstrndup(s, colon - s);
+		error_name = colon + 1;
+
+		found = 0;
+		for (i = 0; i < ARRAY_SIZE(errnos); i++) {
+			if (strcmp(error_name, errnos[i].name) == 0) {
+				ret = errnos[i].number;
+				found = 1;
+				break;
+			}
+		}
+		if (!found)
+			ret = str2num_or_err(
+					colon + 1, 10, _("Unknown errno"), 0, INT_MAX);
+	} else {
+		name = s;
+	}
+
+	found = 0;
+	for (i = 0; i < n_entities; i++) {
+		if (strcmp(name, entities[i].name) == 0) {
+			blocked_number = entities[i].number;
+			found = 1;
+			break;
+		}
+	}
+	if (!found)
+		blocked_number = str2num_or_err(
+				name, 10, _("Unknown syscall"), 0, LONG_MAX);
+
+	blocked = xmalloc(sizeof(*blocked));
+	blocked->number = blocked_number;
+	blocked->ret = ret;
+
```

Comment:
Maybe we can make some static analyzer more happy there by:
```
if (name != s)
  free(name);
```

Reply:
Done, needed a cast though.